### PR TITLE
Upgrading to new Key Vault (and MSAL) libraries

### PR DESCRIPTION
**Pull request by srnagar**
This PR migrates from the previous-generation Azure Key Vault library and ADAL libraries to the more modern Azure Key Vault and MSAL equivalents.

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -38,8 +38,8 @@ jobs:
...
--DapplicationClientID=$(applicationClientID) -DapplicationKey=$(applicationKey) -DkeyID=$(keyID) -DwindowsKeyPath=$(windowsKeyPath) -DenclaveAttestationUrl=$(enclaveAttestationUrl) -DenclaveAttestationProtocol=$(enclaveAttestationProtocol) -DenclaveServer=$(enclaveServer)'
+      goals: 'clean dependency:purge-local-repository -Dmssql_jdbc_test_connection_properties=jdbc:sqlserver://$(Target_SQL)$(server_domain);$(database);$(user);$(password); install -Pjre14 -DuserNTLM=$(userNTLM) -DpasswordNTLM=$(passwordNTLM) -DdomainNTLM=$(domainNTLM) -DexcludedGroups=$(Ex_Groups) -Dpkcs12_truststore_password=$(pkcs12_truststore_password) -Dpkcs12_truststore=$(pkcs12_truststore.secureFilePath)
+-DapplicationClientID=$(applicationClientID) -DapplicationKey=$(applicationKey) -DkeyID=$(keyID) -DwindowsKeyPath=$(windowsKeyPath) -DenclaveAttestationUrl=$(enclaveAttestationUrl) -DenclaveAttestationProtocol=$(enclaveAttestationProtocol) -DenclaveServer=$(enclaveServer) -DtenantID=$(tenantID)'
```
Comment (lilgreenbird) [inline azure-pipelines.yml:42]: I don't think we need any changes to this file anymore since tenantID is gone

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 2]
```diff
@@ -669,6 +669,8 @@ boolean getSendTemporalDataTypesAsStringForBulkCopy() {
     String keyStoreLocation = null;
     String keyStorePrincipalId = null;
 
+    String keyVaultProviderTenantId = null;
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:672]: remove, no longer needed

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java [hunk 3]
```diff
@@ -49,668 +53,682 @@
...
-        if (signatureLength != keySizeInBytes) {
-            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_AKVSignatureLengthError"));
-            Object[] msgArgs = {signatureLength, keySizeInBytes, masterKeyPath};
-            throw new SQLServerException(this, form.format(msgArgs), null, 0, false);
+        /**
+         * Constructs a SQLServerColumnEncryptionAzureKeyVaultProvider using the provided TokenCredential to
+         * authenticate to AAD. This is used by KeyVaultClient at runtime to authenticate to Azure Key Vault.
+         *
+         * @param tokenCredential The TokenCredential to use to authenticate to Azure Key Vault.
+         */
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java:144]: [WARNING] Javadoc Warnings
[WARNING] C:\mssql-jdbc\src\main\java\com\microsoft\sqlserver\jdbc\SQLServerColumnEncryptionAzureKeyVaultProvider.java:145: warning: no @throws for com.microsoft.sqlserver.jdbc.SQLServerException
[WARNING] public SQLServerColumnEncryptionAzureKeyVaultProvider(TokenCredential tokenCredential)
[WARNING] ^

File: pom.xml [hunk 4]
```diff
@@ -60,9 +60,8 @@
 		<releaseExt>-preview</releaseExt>
 
 		<!-- Driver Dependencies -->
-		<azure.keyvault.version>1.2.4</azure.keyvault.version>
-		<azure.adal4j.version>1.6.5</azure.adal4j.version>
-		<rest.client.version>1.7.4</rest.client.version>
+		<azure.keyvault.version>4.1.4</azure.keyvault.version>
```
Comment (ulvii) [inline pom.xml:63]: Could you confirm that these are production release versions?

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCredential.java [hunk 5]
```diff
@@ -5,86 +5,118 @@
...
+    /**
+     * Creates a KeyVaultCredential with the given identity client options.
+     *
+     * @param clientId
+     *        the client ID of the application
+     * @param clientSecret
+     *        the secret value of the AAD application.
+     */
+    KeyVaultCredential(String clientId, String clientSecret) {
+        Objects.requireNonNull(clientSecret, "'clientSecret' cannot be null.");
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCredential.java:48]: @lilgreenbird 
This is not how we handle exception messages. Error messages need to be localized and also make sure to throw SQLServerException. Please make sure to change this in all occurrences.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCredential.java [hunk 6]
```diff
@@ -5,86 +5,118 @@
...
+    private final String clientSecret;
+    private String authorization;
+    private ConfidentialClientApplication confidentialClientApplication;
 
-    KeyVaultCredential(String clientId) throws SQLServerException {
+    /**
+     * Creates a KeyVaultCredential with the given identity client options.
+     *
+     * @param clientId
+     *        the client ID of the application
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCredential.java:43]: . at the end.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCredential.java [hunk 7]
```diff
@@ -5,86 +5,118 @@
...
-                    throw new RuntimeException(e);
-                }
-            } else {
-                AuthenticationResult token = getAccessTokenFromClientCredentials(authorization, resource, clientId,
-                        clientKey);
-                accessToken = token.getAccessToken();
-            }
+    private ConfidentialClientApplication getConfidentialClientApplication() {
+        if (null == clientId) {
+            throw logger.logExceptionAsError(new IllegalArgumentException(
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCredential.java:71]: This is not how exception are handled in the driver, @lilgreenbird please make sure to follow the guidelines.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCredential.java [hunk 8]
```diff
@@ -5,86 +5,118 @@
...
+        this.clientSecret = clientSecret;
     }
 
-    KeyVaultCredential() {}
-
-    KeyVaultCredential(String clientId, String clientKey) {
-        this.clientId = clientId;
-        this.clientKey = clientKey;
+    @Override
+    public Mono<AccessToken> getToken(TokenRequestContext request) {
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCredential.java:55]: Please pay attention  to public APIs. We don't want to create public APIs unless it is by design.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/JDBCEncryptionDecryptionTest.java [hunk 9]
```diff
@@ -134,10 +126,10 @@ public void testBadAkv(String serverName, String url, String protocol) throws Ex
 
         try {
             SQLServerColumnEncryptionAzureKeyVaultProvider akv = new SQLServerColumnEncryptionAzureKeyVaultProvider(
-                    (SQLServerKeyVaultAuthenticationCallback) null);
+                    (TokenCredential) null);
             fail(TestResource.getResource("R_expectedExceptionNotThrown"));
-        } catch (SQLServerException e) {
-            assertTrue(e.getMessage().matches(TestUtils.formatErrorMsg("R_NullValue")));
+        } catch (NullPointerException exception) {
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/JDBCEncryptionDecryptionTest.java:131]: Driver should not throw a NullPointerException . Better add a null check to the constructor and throw SQLServerException .

File: src/main/java/com/microsoft/sqlserver/jdbc/ScopeTokenCache.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
+package com.microsoft.sqlserver.jdbc;
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/ScopeTokenCache.java:1]: Need to add license header here.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java [hunk 11]
```diff
@@ -198,7 +182,7 @@ public SQLServerColumnEncryptionAzureKeyVaultProvider(
         }
 
         // Validate encryptionAlgorithm
-        encryptionAlgorithm = this.validateEncryptionAlgorithm(encryptionAlgorithm);
+        KeyWrapAlgorithm _encryptionAlgorithm = this.validateEncryptionAlgorithm(encryptionAlgorithm);
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java:185]: CamelCase here

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultHttpPipelineBuilder.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+import com.azure.core.http.policy.RetryPolicy;
+import com.azure.core.http.policy.UserAgentPolicy;
+import com.azure.core.util.Configuration;
+import com.azure.core.util.logging.ClientLogger;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+
+final class KeyVaultHttpPipelineBuilder {
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultHttpPipelineBuilder.java:23]: The methods in this class should probably not be public.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCustomCredentialPolicy.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
+/*
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCustomCredentialPolicy.java:1]: The methods in this class should probably not be public.

File: pom.xml (hunk 4 above)
Comment (srnagar) [reply pom.xml:63]: Yes, these are production release versions but since we had made these changes a while ago, the versions were from a previous release. We now have new releases and I will update the PR to use the latest versions.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultHttpPipelineBuilder.java (hunk 12 above)
Comment (srnagar) [reply src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultHttpPipelineBuilder.java:23]: I will change the visibility of these methods to package-private.

File: build.gradle [hunk 14]
```diff
@@ -127,15 +125,14 @@ dependencies {
...
 			'org.slf4j:slf4j-nop:1.7.29',
 			'org.antlr:antlr4-runtime:4.7.2',
 			'org.eclipse.gemini.blueprint:gemini-blueprint-mock:2.1.0.RELEASE',
 			'com.google.code.gson:gson:2.8.6',
-			'org.bouncycastle:bcprov-jdk15on:1.64',
-			'com.microsoft.azure:adal4j:1.6.4',
-			'com.microsoft.azure:azure-keyvault:1.2.2',
-			'com.microsoft.azure:azure-keyvault-webkey:1.2.1',
+			'org.bouncycastle:bcprov-jdk15on:1.65',
+			'com.azure:azure-security-keyvault-keys:4.2.0',
```
Comment (ulvii) [inline build.gradle:135]: There is a newer version of this library available. 1.1.3

File: build.gradle [hunk 15]
```diff
@@ -110,14 +110,12 @@ repositories {
 dependencies {
 	compile 'org.osgi:org.osgi.core:6.0.0',
 			'org.osgi:org.osgi.compendium:5.0.0'
-	compileOnly 'com.microsoft.azure:azure-keyvault:1.2.2',
-			'com.microsoft.azure:azure-keyvault-webkey:1.2.1',
-			'com.microsoft.rest:client-runtime:1.7.0',
-			'com.microsoft.azure:adal4j:1.6.4',
+	compileOnly 'com.azure:azure-security-keyvault-keys:4.2.0',
```
Comment (ulvii) [inline build.gradle:113]: There is a newer version of this library available: 4.2.1

File: pom.xml [hunk 16]
```diff
@@ -60,9 +60,8 @@
 		<releaseExt>-preview</releaseExt>
 
 		<!-- Driver Dependencies -->
-		<azure.keyvault.version>1.2.4</azure.keyvault.version>
-		<azure.adal4j.version>1.6.5</azure.adal4j.version>
-		<rest.client.version>1.7.4</rest.client.version>
+		<azure.keyvault.version>4.2.1</azure.keyvault.version>
+		<azure.identity.version>1.1.2</azure.identity.version>
```
Comment (ulvii) [inline pom.xml:64]: There is a newer version of azure-identity available. 1.1.3

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java [hunk 17]
```diff
@@ -648,5 +648,6 @@ static String getResource(String key) {
             {"R_clientCertError", "Reading client certificate failed. Please verify the location of the certificate."},
             {"R_unassignableError", "The class specified by the {0} property must be assignable to {1}."},
             {"R_InvalidCSVQuotes",
-                    "Failed to parse the CSV file, verify that the fields are correctly enclosed in double quotes."},};
+                    "Failed to parse the CSV file, verify that the fields are correctly enclosed in double quotes."},
+            {"R_TokenRequireUrl", "Token credentials require a URL using the HTTPS protocol scheme"},};
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java:652]: Can the url be HTTP? Also please add a `.` to the end of the error message.

File: pom.xml [hunk 18]
```diff
@@ -60,9 +60,8 @@
 		<releaseExt>-preview</releaseExt>
 
 		<!-- Driver Dependencies -->
-		<azure.keyvault.version>1.2.4</azure.keyvault.version>
-		<azure.adal4j.version>1.6.5</azure.adal4j.version>
-		<rest.client.version>1.7.4</rest.client.version>
+		<azure.keyvault.version>4.2.1</azure.keyvault.version>
```
Comment (ulvii) [inline pom.xml:63]: When I load the project in Eclipse, I get  dependency conflicts. @lilgreenbird  I thought you had a fix for it, but can't see it on the PR?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java (hunk 17 above)
Comment (ulvii) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java:652]: I figured HTTP is not supported, so the question can be ignored.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCustomCredentialPolicy.java [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+     *
+     * @param context
+     *        The request context
+     * @param next
+     *        The next HTTP pipeline policy to process the {@code context's} request after this policy completes.
+     * @return A {@link Mono} representing the HTTP response that will arrive asynchronously.
+     */
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        if ("http".equals(context.getHttpRequest().getUrl().getProtocol())) {
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCustomCredentialPolicy.java:60]: If only HTTPS supported lets change this to  `if (!"https".equals(context.getHttpRequest().getUrl().getProtocol())) {`

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+    KeyVaultTokenCredential(SQLServerKeyVaultAuthenticationCallback authenticationCallback) {
+        this.authenticationCallback = authenticationCallback;
+        this.clientId = null;
+        this.clientSecret = null;
+    }
+
+    @Override
+    public Mono<AccessToken> getToken(TokenRequestContext request) {
+        if (null != authenticationCallback) {
+            // If the callback is not null, invoke the callback to get the token. This gets invoked each time
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java:83]: Please change this to multi-line comment.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+import java.util.concurrent.CompletableFuture;
+import reactor.core.publisher.Mono;
+
+
+/**
+ * An AAD credential that acquires a token with a client secret for an AAD application.
+ */
+@Immutable
+class KeyVaultTokenCredential implements TokenCredential {
+    private final ClientLogger logger = new ClientLogger(KeyVaultTokenCredential.class);
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java:33]: For consistency, please change this to use the driver's logger.  You can take a look at KerbAuthentication.java for reference usage.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+     * @param request The context for requesting the token including the scope.
+     * @return The cached access token if it's not expired.
+     */
+    private Mono<AccessToken> authenticateWithConfidentialClientCache(TokenRequestContext request) {
+        return Mono.fromFuture(() -> {
+            SilentParameters.SilentParametersBuilder parametersBuilder = SilentParameters
+                    .builder(new HashSet<>(request.getScopes()));
+            try {
+                return confidentialClientApplication.acquireTokenSilently(parametersBuilder.build());
+            } catch (MalformedURLException e) {
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java:159]: Could you explain why we log a warning here instead of an exception?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java [hunk 23]
```diff
@@ -51,29 +56,35 @@
 
     private final static java.util.logging.Logger akvLogger = java.util.logging.Logger
             .getLogger("com.microsoft.sqlserver.jdbc.SQLServerColumnEncryptionAzureKeyVaultProvider");
+    public static final int KEY_NAME_INDEX = 4;
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java:59]: These variables shouldn't be public.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerKeyVaultAuthenticationCallback.java [hunk 24]
```diff
@@ -1,26 +1,26 @@
...
+
+package com.microsoft.sqlserver.jdbc;
+
+/**
+ * Provides a callback delegate which is to be implemented by the client code
+ *
+ */
+public interface SQLServerKeyVaultAuthenticationCallback {
+
+        /**
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerKeyVaultAuthenticationCallback.java:14]: Please revert the changes to this file.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java [hunk 25]
```diff
@@ -124,56 +198,43 @@ public SQLServerColumnEncryptionAzureKeyVaultProvider(
...
      * Constructs a SQLServerColumnEncryptionAzureKeyVaultProvider with a callback function to authenticate to AAD. This
      * is used by KeyVaultClient at runtime to authenticate to Azure Key Vault.
-     * 
+     *
      * @param authenticationCallback
      *        - Callback function used for authenticating to AAD.
      * @throws SQLServerException
      *         when an error occurs
      */
+    @Deprecated
```
Comment (David-Engel) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java:207]: Do we need to deprecate this public interface? If so, why?

I do recall the previous discussion about removing it because we thought we couldn't support it via MSAL. But since we can support it, is there a compelling reason to push applications to the new interface?

Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java:207]: No need to deprecate this constructor anymore, because authenticationCallback works with MSAL now.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java (hunk 22 above)
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java:157]: it doesn't it's actually throwing a RuntimeException and doesn't need to be logged as that' fatal

File: src/test/java/com/microsoft/sqlserver/jdbc/fedauth/FedauthWithAE.java [hunk 26]
```diff
@@ -128,7 +121,7 @@ public void testFedAuthWithAE_AKV() throws SQLException {
             dropCMK(stmt, cmkName3);
             setupCMK_AKVOld(cmkName3, stmt);
 
-            createCEK(cmkName3, setupKeyStoreProvider_AKVOld(), stmt, keyIDs[0]);
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/fedauth/FedauthWithAE.java:131]: This test should not be removed, it is the only test thats actually using `SQLServerColumnEncryptionAzureKeyVaultProvider(authenticationCallback)`.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java (hunk 25 above)
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java]: no it doesn't anymore, I've removed that

File: build.gradle (hunk 14 above)
Comment (ulvii) [inline build.gradle:135]: In pom file azure-security-keyvault-keys version is 4.2.1, its 4.2.0 here.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java (hunk 20 above)
Comment (ulvii) [reply src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java:81]: Not resolved.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MSITest.java [hunk 27]
```diff
@@ -299,7 +300,9 @@ public void testNumericAkvWithBadCred() throws SQLException {
             testNumericAKV(connStr);
             fail(TestResource.getResource("R_expectedFailPassed"));
         } catch (Exception e) {
-            assert (e.getMessage().contains("AuthenticationException"));
+            assertTrue(e.getCause() instanceof MsalServiceException);
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MSITest.java:303]: What is the actual exception type here? is it being thrown as SQLServerException?

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/JDBCEncryptionDecryptionTest.java [hunk 28]
```diff
@@ -91,18 +92,15 @@ public void testJksName(String serverName, String url, String protocol) throws E
      */
     @ParameterizedTest
     @MethodSource("enclaveParams")
+    @Tag(Constants.reqExternalSetup)
     public void testAkvName(String serverName, String url, String protocol) throws Exception {
         setAEConnectionString(serverName, url, protocol);
 
-        try {
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/JDBCEncryptionDecryptionTest.java:97]: Please do not remove this test, it is still valid.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java [hunk 29]
```diff
@@ -84,96 +95,126 @@ public String getName() {
...
+     * @param tokenCredential
+     *        The TokenCredential to use to authenticate to Azure Key Vault.
      * 
      * @throws SQLServerException
      *         when an error occurs
      */
-    SQLServerColumnEncryptionAzureKeyVaultProvider() throws SQLServerException {
-        credentials = new KeyVaultCredential();
-        keyVaultClient = new KeyVaultClient(credentials);
+    public SQLServerColumnEncryptionAzureKeyVaultProvider(TokenCredential tokenCredential) throws SQLServerException {
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java:166]: There is no tests for this constructor, please add a few tests. I could only find one test, which only tests bad AKV name.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCustomCredentialPolicy.java [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+                .map(header -> extractChallenge(header, BEARER_TOKEN_PREFIX)).flatMap(map -> {
+                    keyVaultTokenCredential.setAuthorization(map.get("authorization"));
+                    keyVaultTokenCredential.setResource(map.get("resource"));
+                    keyVaultTokenCredential.setScope(map.get("scope"));
+                    cache.setRequest(new TokenRequestContext().addScopes(map.get("resource") + "/.default"));
+                    return cache.getToken();
+                }).flatMap(token -> {
+                    context.getHttpRequest().setHeader(AUTHORIZATION, BEARER_TOKEN_PREFIX + token.getToken());
+                    return next.process();
+                });
```
Comment (rene-ye) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCustomCredentialPolicy.java:76]: This is very interesting syntax but can we either link a source for a spec detailing why we want these specific fields or add some comments to explain. The string field names should also be changed to constants incase it changes in the future.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/JDBCEncryptionDecryptionTest.java [hunk 31]
```diff
@@ -2244,22 +2234,69 @@ void testNumerics(SQLServerStatement stmt, String cekName, String[][] table, Str
         }
     }
 
-    SQLServerKeyVaultAuthenticationCallback authenticationCallback = new SQLServerKeyVaultAuthenticationCallback() {
-        // @Override
-        ExecutorService service = Executors.newFixedThreadPool(2);
+    @ParameterizedTest
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/JDBCEncryptionDecryptionTest.java:2237]: None of these new tests seem to actually test authenticationCallback .As far as I understand, driver applications will be able to implement  `SQLServerKeyVaultAuthenticationCallback.getAccessToken()` API in 2 ways:
- Using adal4j library, see the sample [here](https://docs.microsoft.com/en-us/sql/connect/jdbc/azure-key-vault-sample-version-7.0?view=sql-server-ver15).
- Using msal4j library, as implemented below.
Regardless of which method the application chooses to use, the driver needs to be able to handle both. Please add appropriate tests for both scenarios.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/JDBCEncryptionDecryptionTest.java [hunk 32]
```diff
@@ -134,7 +132,7 @@ public void testBadAkv(String serverName, String url, String protocol) throws Ex
 
         try {
             SQLServerColumnEncryptionAzureKeyVaultProvider akv = new SQLServerColumnEncryptionAzureKeyVaultProvider(
-                    (SQLServerKeyVaultAuthenticationCallback) null);
+                    (TokenCredential) null);
```
Comment (ulvii) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/JDBCEncryptionDecryptionTest.java:135]: Again, we are not removing the SQLServerKeyVaultAuthenticationCallback anymore, so keep this test as it is and add a test for TokenCredential separately.

File: src/test/java/com/microsoft/sqlserver/jdbc/fedauth/FedauthWithAE.java (hunk 26 above)
Comment (peterbae) [reply src/test/java/com/microsoft/sqlserver/jdbc/fedauth/FedauthWithAE.java:131]: SQLServerColumnEncryptionAzureKeyVaultProvider(authenticationCallback) is deprecated for removal in the next stable release, so I think we can remove calling this test method.

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java [hunk 33]
```diff
@@ -0,0 +1,203 @@
...
+    private final String clientId;
+    private final String clientSecret;
+    private final SQLServerKeyVaultAuthenticationCallback authenticationCallback;
+    private String authorization;
+    private ConfidentialClientApplication confidentialClientApplication;
+    private String resource;
+    private String scope;
+
+    /**
+     * Creates a KeyVaultCredential with the given identity client options.
```
Comment (peterbae) [inline src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java:41]: Shouldn't the comment here and below for the constructors say "Creates a KeyVaultTokenCredential" instead of "Creates  a KeyVaultCredential"?

File: src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java (hunk 20 above)
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java:81]: sorry you're right, this change isn't merged to this PR yet..

**Review by David-Engel — COMMENTED** [review]
I just noticed we might be overlooking a scenario in the code changes:
https://github.com/microsoft/mssql-jdbc/pull/1413/files#diff-0a4455f5ffd8e31e9e1badaa7822375eL4454 (Old/Left line 4454 in SQLServerConnection.java)
MSAL should be able to cover the SqlAuthentication.ActiveDirectoryIntegrated scenario, right? Previously the driver had to rely on a native call (ultimately down to adal.dll) that I believe (hope) can now be accomplished via MSAL. We can then eliminate the underlying native call.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java [hunk 34]
```diff
@@ -51,29 +55,35 @@
...
 
-    private KeyVaultClient keyVaultClient;
+    private Map<String, KeyClient> cachedKeyClients = new ConcurrentHashMap<>();
+    private Map<String, CryptographyClient> cachedCryptographyClients = new ConcurrentHashMap<>();
+    private TokenCredential credential;
 
-    private KeyVaultCredential credentials;
+    static {
+        akvTrustedEndpoints = getTrustedEndpoints();
+    }
```
Comment (rene-ye) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java:86]: Can just declare

`private static final List<String> akvTrustedEndpoints = getTrustedEndpoints();` since `getTrustedEndpoints` is static.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java (hunk 17 above)
Comment (ulvii) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java:652]: @lilgreenbird  please add `.` to the end of the error message.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/JDBCEncryptionDecryptionTest.java (hunk 28 above)
Comment (lilgreenbird) [reply src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/JDBCEncryptionDecryptionTest.java:97]: this is not removed, it's been renamed to testAkvNameWithAuthCallback

**Review by peterbae — APPROVED** [review]

File: build.gradle [hunk 35]
```diff
@@ -110,14 +110,12 @@ repositories {
 dependencies {
 	compile 'org.osgi:org.osgi.core:6.0.0',
 			'org.osgi:org.osgi.compendium:5.0.0'
-	compileOnly 'com.microsoft.azure:azure-keyvault:1.2.2',
-			'com.microsoft.azure:azure-keyvault-webkey:1.2.1',
-			'com.microsoft.rest:client-runtime:1.7.0',
-			'com.microsoft.azure:adal4j:1.6.4',
+	compileOnly 'com.azure:azure-security-keyvault-keys:4.2.1',
+			'com.azure:azure-identity:1.1.0',
```
Comment (ulvii) [inline build.gradle:114]: This is supposde to be 1.1.3

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java [hunk 36]
```diff
@@ -404,7 +404,7 @@ static String getResource(String key) {
                     "FEDAUTHINFO token stream is not long enough ({0}) to contain the data it claims to."},
             {"R_FedAuthInfoDoesNotContainStsurlAndSpn",
                     "FEDAUTHINFO token stream does not contain both STSURL and SPN."},
-            {"R_ADALExecution", "Failed to authenticate the user {0} in Active Directory (Authentication={1})."},
+            {"R_MSALExecution", "Failed to authenticate the user {0} in Active Directory (Authentication={1})."},
```
Comment (ulvii) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java:407]: Please search for `ADAL` in SQLServerResource and make similar changes to the error messages that mentioned ADAL.

**Review by ulvii — APPROVED** [review]

**Review by rene-ye — APPROVED** [review]

**Review by peterbae — APPROVED** [review]

**Review by lilgreenbird — APPROVED** [review]
